File: java/main/java/com/github/mrr736/fetchiso/Downloader.java
/*
 * Downloader for Modern Java
 * version 1.1
 * https://github.com/MrR736/FetchISO
 *
 * SPDX-FileCopyrightText: 2026 MrR736
 * SPDX-License-Identifier: MIT
 */

package com.github.mrr736.fetchiso;

import java.io.*;
import java.nio.file.*;

public class Downloader {
	private volatile String lastError = "";

	public boolean downloadFile(String url,String file) {
		ProcessBuilder builder;
		try {
			if (isCommandAvailable("wget")) {
				builder = new ProcessBuilder("wget","-q","--show-progress","-O",file,url);
			} else if (isCommandAvailable("curl")) {
				builder = new ProcessBuilder("curl","-L","--progress-bar","-o",file,url);
			} else if (isWindows() && isCommandAvailable("powershell")) {
				builder = new ProcessBuilder(
					"powershell", "-Command",
					"Invoke-WebRequest -Uri \"" + url +
					"\" -OutFile \"" + file + "\""
				);

			} else {
				lastError = "No downloader found (wget, curl, powershell)";
				return false;
			}

			Process process = builder.redirectErrorStream(true).start();
			String output = readOutput(process);

			int exit = process.waitFor();
			if (exit != 0) {
				lastError = "Downloader failed: " + output;
				return false;
			}

			if (!Files.exists(Paths.get(file))) {
				lastError = "File not created: " + file;
				return false;
			}
			return true;
		} catch (Exception e) {
			lastError = "Download error: " + e.getMessage();
			return false;
		}
	}

	public String getLastError() {
		return lastError;
	}

	private boolean isCommandAvailable(String command) {
		try {
			Process process;

			if (isWindows()) {
				process = new ProcessBuilder("where",command).start();
			} else {
				process = new ProcessBuilder("sh","-c","command -v " + command).start();
			}

			return process.waitFor() == 0;
		} catch (Exception e) {
			return false;
		}
	}

	private String readOutput(Process process) {
		StringBuilder out = new StringBuilder();

		try (
			BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))
		) {

			String line;
			while ((line = reader.readLine()) != null) {
				out.append(line).append('\n');
			}
		} catch (IOException e) {
			out.append(e.getMessage());
		}

		return out.toString();
	}

	private boolean isWindows() {
		return System.getProperty("os.name").toLowerCase().contains("win");
	}
}

File: python/downloader.py
"""/*
// Downloader for Modern Python
// version 1.0
// https://github.com/MrR736/FetchISO
//
// SPDX-FileCopyrightText: 2025 MrR736 <https://github.com/MrR736>
// SPDX-License-Identifier: MIT
*/"""

import os
import subprocess
from typing import Optional


class Downloader:
    def __init__(self):
        self.last_error: Optional[str] = None

    def download_file(self, url: str, file: str) -> bool:
        """Downloads a file using wget or curl."""
        if self.is_command_available("wget"):
            command = f"wget -q --show-progress -O \"{file}\" \"{url}\""
        elif self.is_command_available("curl"):
            command = f"curl -L --progress-bar -o \"{file}\" \"{url}\""
        else:
            self.last_error = "No suitable download tool found (wget or curl)"
            return False

        # Execute command and capture output
        ret_code, output = self.execute_command(command)

        if ret_code != 0:
            self.last_error = f"Download failed with exit code: {ret_code}\n{output}"
            return False

        # Ensure the file actually exists after download
        if not os.path.exists(file):
            self.last_error = f"Download completed but file does not exist: {file}"
            return False

        return True

    def get_last_error(self) -> Optional[str]:
        """Returns the last error message."""
        return self.last_error

    @staticmethod
    def is_command_available(cmd: str) -> bool:
        """Checks if a command is available on the system."""
        check_cmd = "where" if os.name == "nt" else "command -v"
        return subprocess.call(f"{check_cmd} {cmd}", shell=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL) == 0

    @staticmethod
    def execute_command(command: str):
        """Executes a shell command and returns (exit_code, output)."""
        try:
            result = subprocess.run(command, shell=True, check=False, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True)
            return result.returncode, result.stdout.strip()
        except Exception as e:
            return -1, str(e)
